### ISSUE TITLE
Possibly wrong error: redeclaration of 'x' with a different type

### ISSUE THREAD

**Issue by pmor13** [llvm/llvm-project#57480]
This C code:
```
static int (*x)[];

void f1(void)
{
    extern int (*x)[3];
}

void f2(void)
{
    extern int (*x)[5];
}
```
leads to:
```
error: redeclaration of 'x' with a different type: 'int (*)[5]' vs 'int (*)[3]'
    extern int (*x)[5];
                 ^
<source>:5:18: note: previous declaration is here
    extern int (*x)[3];
```
which is unexpected because the previous declaration is invisible. Hence, there is no need to check for type compatibility.

Relevant quote from C11, 6.2.2 Linkages of identifiers, 4:
> For an **identifier declared with the storage-class specifier extern in a scope in which a
prior declaration of that identifier is visible**,31) if the prior declaration specifies internal or
external linkage, the linkage of the identifier at the later declaration is the same as the
linkage specified at the prior declaration. If no prior declaration is visible, or if the prior
declaration specifies no linkage, then the identifier has external linkage.

For identifier `x` in function `f2` declaration of identifier `x` in function `f1` is invisible. Hence, there is no need to check for type compatibility. I see it this way. C experts are welcome. GCC generates no diagnostics (even under `-std=c11 -pedantic -Wall -Wextra -Wno-unused-variable`).

**Comment by AaronBallman** [comment]
This is a situation where the standard does not require a diagnostic (it violates a semantic restriction, not a constraint), but Clang elects to give you one anyway because the code is incorrect. The diagnostic was introduced in https://lists.llvm.org/pipermail/cfe-commits/Week-of-Mon-20130812/085782.html.

Specifying that a variable is `extern` means "you'll find this definition elsewhere", so when two differently scoped `extern` declarations resolve to the same definition but have incompatible types, one of those two declarations is wrong, which is what this diagnostic is trying to tell you.

So I don't think this is a bug per:

C2x 6.2.2p2: "In the set of translation units and libraries that constitutes an entire program, each declaration of a particular identifier with external linkage denotes the same object or function."

C2x 6.2.7p2: "All declarations that refer to the same object or function shall have compatible type; otherwise, the behavior is undefined."

Closing as not a bug, but if you still feel there is an issue here, please reopen with more details.

**Comment by pmor13** [comment]
@AaronBallman Indeed, per C2x 6.2.7p2 the code above triggers UB. Thanks. TIL.

**Comment by AaronBallman** [comment]
> @AaronBallman Indeed, per C2x 6.2.7p2 the code above triggers UB. Thanks. TIL.

It does, but that UB isn't the reason for the behavior you're seeing. You'd get the same behavior if the file scope declaration was also `extern`: https://godbolt.org/z/aq6Yjjnsn

The key bit is that all external declarations to the same identifier need to refer to the same object or function, which means that the types of all those external declarations need to be compatible (since the object or function can only have one type at the end of the day).

**Comment by pmor13** [comment]
@AaronBallman FYI: GCC generates no diagnostics because it is [17 yo regression](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=21343).

**Comment by pmor13** [comment]
@AaronBallman FYI: Clang doesn't diagnose whether the same identifier appears with both internal and external linkage: https://godbolt.org/z/8KaPPT38a. Here is a [dedicated issue](https://github.com/llvm/llvm-project/issues/54215).
